File: mac-tracker/Tests/WorkTrackerTrackerTests/ActivityQueueTests.swift
import XCTest
@testable import WorkTrackerTracker

private final class FakeEventsAPIClient: EventsAPIClient {
    var shouldFail = false
    var failure: APIClientError = .requestFailed(statusCode: 500)
    /// Runs while a request is "in flight", to simulate activity arriving mid-flush.
    var onPost: (() -> Void)?
    private(set) var receivedBatches: [[Date]] = []

    func postEvents(_ timestamps: [Date], serverBaseURL: String, apiKey: String) async throws {
        receivedBatches.append(timestamps)
        onPost?()
        if shouldFail {
            throw failure
        }
    }
}

/// A controllable clock so backoff/debounce timing can be tested without
/// real `Timer`s or `sleep`s — mirrors how `SystemIdleTimeSource` is faked
/// for `IdleMonitor` in IdleMonitorTests.swift.
private final class FakeClock {
    var current: Date
    init(_ current: Date = Date(timeIntervalSince1970: 0)) { self.current = current }
    func advance(by seconds: TimeInterval) { current = current.addingTimeInterval(seconds) }
    func now() -> Date { current }
}

final class ActivityQueueTests: XCTestCase {
    private var tempURL: URL!

    override func setUp() {
        super.setUp()
        tempURL = FileManager.default.temporaryDirectory
            .appendingPathComponent(UUID().uuidString)
            .appendingPathComponent("queue.json")
    }

    override func tearDown() {
        try? FileManager.default.removeItem(at: tempURL.deletingLastPathComponent())
        super.tearDown()
    }

    // MARK: - Basic enqueue/flush behavior

    func test_newQueue_startsEmpty() {
        let queue = ActivityQueue(storageURL: tempURL)
        XCTAssertEqual(queue.pendingCount, 0)
    }

    func test_enqueue_incrementsPendingCount() {
        let queue = ActivityQueue(storageURL: tempURL)
        queue.enqueue(Date())
        queue.enqueue(Date())
        XCTAssertEqual(queue.pendingCount, 2)
    }

    func test_flush_onSuccess_clearsTheQueueAndSendsOneBatch() async {
        let queue = ActivityQueue(storageURL: tempURL)
        queue.enqueue(Date())
        queue.enqueue(Date())
        let client = FakeEventsAPIClient()

        await queue.flush(client: client, serverBaseURL: "https://example.vercel.app", apiKey: "wtk_live_x")

        XCTAssertEqual(queue.pendingCount, 0)
        XCTAssertEqual(client.receivedBatches.count, 1)
        XCTAssertEqual(client.receivedBatches[0].count, 2)
    }

    func test_flush_onFailure_keepsEventsQueuedForRetry() async {
        let queue = ActivityQueue(storageURL: tempURL)
        queue.enqueue(Date())
        let client = FakeEventsAPIClient()
        client.shouldFail = true

        await queue.flush(client: client, serverBaseURL: "https://example.vercel.app", apiKey: "wtk_live_x")

        XCTAssertEqual(queue.pendingCount, 1)
    }

    // MARK: - lastSuccessfulSyncAt

    func test_lastSuccessfulSyncAt_isNilInitially() {
        let queue = ActivityQueue(storageURL: tempURL)
        XCTAssertNil(queue.lastSuccessfulSyncAt)
    }

    func test_flush_onSuccess_setsLastSuccessfulSyncAt() async {
        let clock = FakeClock()
        let queue = ActivityQueue(storageURL: tempURL, now: clock.now)
        queue.enqueue(clock.now())
        clock.advance(by: 42)

        await queue.flush(client: FakeEventsAPIClient(), serverBaseURL: "https://example.vercel.app", apiKey: "k")

        XCTAssertEqual(queue.lastSuccessfulSyncAt, clock.now())
    }

    func test_flush_onFailure_doesNotSetLastSuccessfulSyncAt() async {
        let queue = ActivityQueue(storageURL: tempURL)
        queue.enqueue(Date())
        let client = FakeEventsAPIClient()
        client.shouldFail = true

        await queue.flush(client: client, serverBaseURL: "https://example.vercel.app", apiKey: "k")

        XCTAssertNil(queue.lastSuccessfulSyncAt)
    }

    func test_lastSuccessfulSyncAt_survivesRestart_byPersistingToDisk() async {
        // A whole-second FakeClock value, not the real system clock: ISO
        // 8601 round-tripping only preserves millisecond precision, so
        // comparing against `Date()`'s full (sub-millisecond) precision
        // would be a flaky equality check.
        let clock = FakeClock()
        let firstInstance = ActivityQueue(storageURL: tempURL, persistDebounceInterval: 0, now: clock.now)
        firstInstance.enqueue(clock.now())
        await firstInstance.flush(client: FakeEventsAPIClient(), serverBaseURL: "https://example.vercel.app", apiKey: "k")

        let secondInstance = ActivityQueue(storageURL: tempURL)
        XCTAssertEqual(secondInstance.lastSuccessfulSyncAt, clock.now())
    }

    func test_loadsAnOlderQueueFile_writtenBeforeLastSuccessfulSyncAtExisted() throws {
        // Pre-upgrade queue files are a bare `[String]` of ISO 8601
        // timestamps, with no lastSuccessfulSyncAt field at all.
        try FileManager.default.createDirectory(at: tempURL.deletingLastPathComponent(), withIntermediateDirectories: true)
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        let legacyJSON = try JSONEncoder().encode([formatter.string(from: Date())])
        try legacyJSON.write(to: tempURL)

        let queue = ActivityQueue(storageURL: tempURL)

        XCTAssertEqual(queue.pendingCount, 1, "the legacy pending entry should still load")
        XCTAssertNil(queue.lastSuccessfulSyncAt, "an old file has no sync history to report")
    }

    func test_flush_onEmptyQueue_doesNotCallTheClient() async {
        let queue = ActivityQueue(storageURL: tempURL)
        let client = FakeEventsAPIClient()

        await queue.flush(client: client, serverBaseURL: "https://example.vercel.app", apiKey: "wtk_live_x")

        XCTAssertEqual(client.receivedBatches.count, 0)
    }

    // MARK: - Persistence round-trip (debounce disabled so writes are synchronous)

    func test_queueSurvivesRestart_byPersistingToDisk() {
        let firstInstance = ActivityQueue(storageURL: tempURL, persistDebounceInterval: 0)
        firstInstance.enqueue(Date())
        firstInstance.enqueue(Date())

        let secondInstance = ActivityQueue(storageURL: tempURL)
        XCTAssertEqual(secondInstance.pendingCount, 2)
    }

    func test_afterSuccessfulFlush_aFreshInstanceHasNothingQueued() async {
        let firstInstance = ActivityQueue(storageURL: tempURL, persistDebounceInterval: 0)
        firstInstance.enqueue(Date())
        await firstInstance.flush(client: FakeEventsAPIClient(), serverBaseURL: "https://example.vercel.app", apiKey: "k")

        let secondInstance = ActivityQueue(storageURL: tempURL)
        XCTAssertEqual(secondInstance.pendingCount, 0)
    }

    // MARK: - Persistence debouncing

    func test_enqueue_withinDebounceWindow_doesNotPersistImmediately() {
        let clock = FakeClock()
        let firstInstance = ActivityQueue(
            storageURL: tempURL, persistDebounceInterval: 5, now: clock.now
        )
        firstInstance.enqueue(Date()) // first write always persists (no prior persist yet)
        clock.advance(by: 1) // still well within the 5s debounce window
        firstInstance.enqueue(Date())

        let secondInstance = ActivityQueue(storageURL: tempURL)
        XCTAssertEqual(secondInstance.pendingCount, 1, "the second enqueue should still be debounced, not yet on disk")
    }

    func test_enqueue_afterDebounceWindowElapses_persists() {
        let clock = FakeClock()
        let firstInstance = ActivityQueue(
            storageURL: tempURL, persistDebounceInterval: 5, now: clock.now
        )
        firstInstance.enqueue(Date())
        clock.advance(by: 6) // past the debounce window
        firstInstance.enqueue(Date())

        let secondInstance = ActivityQueue(storageURL: tempURL)
        XCTAssertEqual(secondInstance.pendingCount, 2)
    }

    func test_persistNow_flushesDebouncedChangesImmediately() {
        let clock = FakeClock()
        let firstInstance = ActivityQueue(
            storageURL: tempURL, persistDebounceInterval: 5, now: clock.now
        )
        firstInstance.enqueue(Date())
        clock.advance(by: 1)
        firstInstance.enqueue(Date()) // debounced, not yet on disk
        firstInstance.persistNow() // e.g. called from applicationWillTerminate

        let secondInstance = ActivityQueue(storageURL: tempURL)
        XCTAssertEqual(secondInstance.pendingCount, 2)
    }

    // MARK: - Queue cap / oldest eviction

    func test_enqueue_beyondCap_dropsOldestEntriesFirst() {
        let queue = ActivityQueue(storageURL: tempURL, maxPendingCount: 3, persistDebounceInterval: 0)
        let base = Date(timeIntervalSince1970: 0)
        let timestamps = (0..<5).map { base.addingTimeInterval(TimeInterval($0)) }

        for timestamp in timestamps {
            queue.enqueue(timestamp)
        }

        XCTAssertEqual(queue.pendingCount, 3)
        // The two oldest (index 0 and 1) should have been evicted, keeping
        // the three most recent.
        XCTAssertEqual(queue.pendingTimestamps, Array(timestamps.suffix(3)))
    }

    func test_enqueue_beyondCap_survivesRestartWithOnlyNewestEntries() {
        let firstInstance = ActivityQueue(storageURL: tempURL, maxPendingCount: 2, persistDebounceInterval: 0)
        let base = Date(timeIntervalSince1970: 0)
        firstInstance.enqueue(base)
        firstInstance.enqueue(base.addingTimeInterval(1))
        firstInstance.enqueue(base.addingTimeInterval(2))

        let secondInstance = ActivityQueue(storageURL: tempURL)
        XCTAssertEqual(secondInstance.pendingCount, 2)
        XCTAssertEqual(secondInstance.pendingTimestamps, [base.addingTimeInterval(1), base.addingTimeInterval(2)])
    }

    // MARK: - Flush backoff on repeated failures

    func test_flush_afterFailure_skipsRetryUntilBackoffElapses() async {
        let clock = FakeClock()
        let queue = ActivityQueue(
            storageURL: tempURL, baseBackoffInterval: 10, maxBackoffInterval: 300, now: clock.now
        )
        queue.enqueue(Date())
        let client = FakeEventsAPIClient()
        client.shouldFail = true

        await queue.flush(client: client, serverBaseURL: "https://example.vercel.app", apiKey: "k")
        XCTAssertEqual(client.receivedBatches.count, 1)
        XCTAssertEqual(queue.currentConsecutiveFailureCount, 1)

        // A retry attempted before the backoff window elapses should be a
        // no-op — the client must not be called again yet.
        clock.advance(by: 5)
        await queue.flush(client: client, serverBaseURL: "https://example.vercel.app", apiKey: "k")
        XCTAssertEqual(client.receivedBatches.count, 1, "should still be backing off")

        // Once the backoff window elapses, the next flush should retry.
        clock.advance(by: 6) // total 11s since the failure, past the 10s base backoff
        await queue.flush(client: client, serverBaseURL: "https://example.vercel.app", apiKey: "k")
        XCTAssertEqual(client.receivedBatches.count, 2)
        XCTAssertEqual(queue.currentConsecutiveFailureCount, 2)
    }

    func test_flush_backoff_doublesWithEachConsecutiveFailureUpToCap() {
        let base: TimeInterval = 10
        let max: TimeInterval = 100
        XCTAssertEqual(ActivityQueue.backoffInterval(forConsecutiveFailures: 1, base: base, max: max), 10)
        XCTAssertEqual(ActivityQueue.backoffInterval(forConsecutiveFailures: 2, base: base, max: max), 20)
        XCTAssertEqual(ActivityQueue.backoffInterval(forConsecutiveFailures: 3, base: base, max: max), 40)
        XCTAssertEqual(ActivityQueue.backoffInterval(forConsecutiveFailures: 4, base: base, max: max), 80)
        // Would be 160 uncapped; the max caps it.
        XCTAssertEqual(ActivityQueue.backoffInterval(forConsecutiveFailures: 5, base: base, max: max), 100)
        XCTAssertEqual(ActivityQueue.backoffInterval(forConsecutiveFailures: 6, base: base, max: max), 100)
    }

    func test_flush_onSuccess_resetsBackoffState() async {
        let clock = FakeClock()
        let queue = ActivityQueue(
            storageURL: tempURL, baseBackoffInterval: 10, maxBackoffInterval: 300, now: clock.now
        )
        queue.enqueue(Date())
        let client = FakeEventsAPIClient()
        client.shouldFail = true
        await queue.flush(client: client, serverBaseURL: "https://example.vercel.app", apiKey: "k")
        XCTAssertEqual(queue.currentConsecutiveFailureCount, 1)

        client.shouldFail = false
        clock.advance(by: 20) // past the backoff window so this attempt actually runs
        queue.enqueue(Date())
        await queue.flush(client: client, serverBaseURL: "https://example.vercel.app", apiKey: "k")

        XCTAssertEqual(queue.currentConsecutiveFailureCount, 0)
        XCTAssertEqual(queue.pendingCount, 0)

        // A subsequent failure should back off starting from the base
        // interval again, not continue escalating from before the reset.
        client.shouldFail = true
        queue.enqueue(Date())
        await queue.flush(client: client, serverBaseURL: "https://example.vercel.app", apiKey: "k")
        XCTAssertEqual(queue.currentConsecutiveFailureCount, 1)
    }

    // MARK: - Outage resilience

    private let url = "https://example.vercel.app"

    func test_flush_sendsLargeBacklogInChunks_andClearsAll() async {
        let queue = ActivityQueue(storageURL: tempURL, chunkSize: 100, persistDebounceInterval: 0)
        for i in 0..<250 { queue.enqueue(Date(timeIntervalSince1970: TimeInterval(i))) }
        let client = FakeEventsAPIClient()

        await queue.flush(client: client, serverBaseURL: url, apiKey: "k")

        XCTAssertEqual(client.receivedBatches.map(\.count), [100, 100, 50])
        XCTAssertEqual(queue.pendingCount, 0)
    }

    func test_flush_failureMidBacklog_keepsOnlyUnsentChunks() async {
        let queue = ActivityQueue(storageURL: tempURL, chunkSize: 100, persistDebounceInterval: 0)
        for i in 0..<250 { queue.enqueue(Date(timeIntervalSince1970: TimeInterval(i))) }
        let client = FakeEventsAPIClient()
        var calls = 0
        client.onPost = { calls += 1; client.shouldFail = calls >= 2 }

        await queue.flush(client: client, serverBaseURL: url, apiKey: "k")

        XCTAssertEqual(queue.pendingCount, 150, "first chunk acknowledged, the rest stay queued")
        XCTAssertEqual(queue.pendingTimestamps.first, Date(timeIntervalSince1970: 100))
    }

    func test_flush_keepsEventsEnqueuedWhileRequestIsInFlight() async {
        let queue = ActivityQueue(storageURL: tempURL, persistDebounceInterval: 0)
        queue.enqueue(Date(timeIntervalSince1970: 1))
        let client = FakeEventsAPIClient()
        var injected = false
        client.onPost = {
            guard !injected else { return }
            injected = true
            queue.enqueue(Date(timeIntervalSince1970: 2))
        }

        await queue.flush(client: client, serverBaseURL: url, apiKey: "k")

        XCTAssertEqual(client.receivedBatches.first, [Date(timeIntervalSince1970: 1)])
        // The late arrival is sent by the same flush's next chunk, not lost.
        XCTAssertEqual(client.receivedBatches.flatMap { $0 }.count, 2)
        XCTAssertEqual(queue.pendingCount, 0)
    }

    func test_flush_capEvictionDuringFlight_doesNotRemoveUnsentEntries() async {
        let queue = ActivityQueue(storageURL: tempURL, maxPendingCount: 3, persistDebounceInterval: 0)
        for i in 0..<3 { queue.enqueue(Date(timeIntervalSince1970: TimeInterval(i))) }
        let client = FakeEventsAPIClient()
        // The post covers entries 0..2, but 0 and 1 are evicted meanwhile:
        // only entry 2 may be removed; 3 and 4 must stay and be sent next.
        var first = true
        client.onPost = {
            guard first else { return }
            first = false
            queue.enqueue(Date(timeIntervalSince1970: 3))
            queue.enqueue(Date(timeIntervalSince1970: 4))
        }

        await queue.flush(client: client, serverBaseURL: url, apiKey: "k")

        XCTAssertEqual(queue.pendingCount, 0)
        XCTAssertEqual(client.receivedBatches.flatMap { $0 }.last, Date(timeIntervalSince1970: 4))
    }

    func test_flush_permanentRejection_dropsChunkAndContinues() async {
        let queue = ActivityQueue(storageURL: tempURL, chunkSize: 2, persistDebounceInterval: 0)
        for i in 0..<4 { queue.enqueue(Date(timeIntervalSince1970: TimeInterval(i))) }
        let client = FakeEventsAPIClient()
        var calls = 0
        client.failure = .requestFailed(statusCode: 400)
        client.onPost = { calls += 1; client.shouldFail = calls == 1 }

        await queue.flush(client: client, serverBaseURL: url, apiKey: "k")

        XCTAssertEqual(queue.pendingCount, 0, "poison chunk dropped, next chunk still delivered")
        XCTAssertEqual(queue.droppedEventCount, 2)
        XCTAssertEqual(client.receivedBatches.count, 2)
    }

    func test_flush_unauthorized_backsOffAndReportsError() async {
        let queue = ActivityQueue(storageURL: tempURL)
        queue.enqueue(Date())
        let client = FakeEventsAPIClient()
        client.shouldFail = true
        client.failure = .unauthorized

        await queue.flush(client: client, serverBaseURL: url, apiKey: "k")

        XCTAssertEqual(queue.pendingCount, 1, "never drop data because of a bad key")
        XCTAssertEqual(queue.lastError, "API key invalid or revoked")
        XCTAssertEqual(queue.currentConsecutiveFailureCount, 1)
    }

    func test_flush_success_clearsLastError() async {
        let clock = FakeClock()
        let queue = ActivityQueue(storageURL: tempURL, baseBackoffInterval: 10, now: clock.now)
        queue.enqueue(Date())
        let client = FakeEventsAPIClient()
        client.shouldFail = true
        await queue.flush(client: client, serverBaseURL: url, apiKey: "k")
        XCTAssertNotNil(queue.lastError)

        client.shouldFail = false
        clock.advance(by: 20)
        await queue.flush(client: client, serverBaseURL: url, apiKey: "k")

        XCTAssertNil(queue.lastError)
    }

    func test_enqueue_beyondCap_countsDroppedEvents() {
        let queue = ActivityQueue(storageURL: tempURL, maxPendingCount: 2, persistDebounceInterval: 0)
        for i in 0..<5 { queue.enqueue(Date(timeIntervalSince1970: TimeInterval(i))) }
        XCTAssertEqual(queue.droppedEventCount, 3)
    }

    func test_load_corruptQueueFile_isQuarantinedNotOverwritten() throws {
        try FileManager.default.createDirectory(at: tempURL.deletingLastPathComponent(), withIntermediateDirectories: true)
        try Data("{not json".utf8).write(to: tempURL)

        let queue = ActivityQueue(storageURL: tempURL)
        queue.enqueue(Date())
        queue.persistNow()

        XCTAssertEqual(queue.pendingCount, 1)
        let quarantined = tempURL.appendingPathExtension("corrupt")
        XCTAssertEqual(try String(contentsOf: quarantined, encoding: .utf8), "{not json")
    }

    // MARK: - Backward compatibility

    func test_existingQueueFileWith3000Events_loadsAndIsFullyDeliveredInChunks() async throws {
        // The shape the previous version wrote: {"pending": [...], "lastSuccessfulSyncAt": "..."}.
        try FileManager.default.createDirectory(at: tempURL.deletingLastPathComponent(), withIntermediateDirectories: true)
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        let stamps = (0..<3000).map { formatter.string(from: Date(timeIntervalSince1970: 1_700_000_000 + TimeInterval($0 * 30))) }
        let json = try JSONSerialization.data(withJSONObject: ["pending": stamps, "lastSuccessfulSyncAt": "2026-01-01T00:00:00.000Z"])
        try json.write(to: tempURL)

        let queue = ActivityQueue(storageURL: tempURL)
        XCTAssertEqual(queue.pendingCount, 3000)
        XCTAssertNotNil(queue.lastSuccessfulSyncAt)

        let client = FakeEventsAPIClient()
        await queue.flush(client: client, serverBaseURL: url, apiKey: "k")

        XCTAssertEqual(client.receivedBatches.map(\.count).reduce(0, +), 3000)
        XCTAssertTrue(client.receivedBatches.allSatisfy { $0.count <= ActivityQueue.defaultChunkSize })
        XCTAssertEqual(queue.pendingCount, 0)
    }
}

File: windows-tracker/test/WorkTrackerTracker.Core.Tests/ActivityQueueTests.cs
using WorkTrackerTracker.Core;

namespace WorkTrackerTracker.Core.Tests;

file sealed class FakeEventsApiClient : IEventsApiClient
{
    public bool ShouldFail { get; set; }
    public ApiClientException Failure { get; set; } = new("simulated failure", 500);
    /// <summary>Runs while a request is "in flight", to simulate activity arriving mid-flush.</summary>
    public Action? OnPost { get; set; }
    public List<IReadOnlyList<DateTimeOffset>> ReceivedBatches { get; } = new();

    public Task PostEventsAsync(IReadOnlyList<DateTimeOffset> timestamps, string serverBaseUrl, string apiKey, CancellationToken cancellationToken = default)
    {
        ReceivedBatches.Add(timestamps.ToArray());
        OnPost?.Invoke();
        if (ShouldFail)
        {
            throw Failure;
        }
        return Task.CompletedTask;
    }
}

public sealed class ActivityQueueTests : IDisposable
{
    private readonly string _tempFile = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString(), "queue.json");

    public void Dispose()
    {
        var directory = Path.GetDirectoryName(_tempFile);
        if (directory is not null && Directory.Exists(directory))
        {
            Directory.Delete(directory, recursive: true);
        }
    }

    [Fact]
    public void NewQueue_StartsEmpty()
    {
        var queue = new ActivityQueue(_tempFile);
        Assert.Equal(0, queue.PendingCount);
    }

    [Fact]
    public void Enqueue_IncrementsPendingCount()
    {
        var queue = new ActivityQueue(_tempFile);
        queue.Enqueue(DateTimeOffset.UtcNow);
        queue.Enqueue(DateTimeOffset.UtcNow);
        Assert.Equal(2, queue.PendingCount);
    }

    [Fact]
    public async Task Flush_OnSuccess_ClearsTheQueueAndSendsOneBatch()
    {
        var queue = new ActivityQueue(_tempFile);
        queue.Enqueue(DateTimeOffset.UtcNow);
        queue.Enqueue(DateTimeOffset.UtcNow);
        var client = new FakeEventsApiClient();

        await queue.FlushAsync(client, "https://example.vercel.app", "wtk_live_x");

        Assert.Equal(0, queue.PendingCount);
        Assert.Single(client.ReceivedBatches);
        Assert.Equal(2, client.ReceivedBatches[0].Count);
    }

    [Fact]
    public async Task Flush_OnFailure_KeepsEventsQueuedForRetry()
    {
        var queue = new ActivityQueue(_tempFile);
        queue.Enqueue(DateTimeOffset.UtcNow);
        var client = new FakeEventsApiClient { ShouldFail = true };

        await queue.FlushAsync(client, "https://example.vercel.app", "wtk_live_x");

        Assert.Equal(1, queue.PendingCount);
    }

    [Fact]
    public async Task Flush_OnEmptyQueue_DoesNotCallTheClient()
    {
        var queue = new ActivityQueue(_tempFile);
        var client = new FakeEventsApiClient();

        await queue.FlushAsync(client, "https://example.vercel.app", "wtk_live_x");

        Assert.Empty(client.ReceivedBatches);
    }

    [Fact]
    public void QueueSurvivesRestart_ByPersistingToDisk()
    {
        var first = new ActivityQueue(_tempFile, persistDebounce: TimeSpan.Zero);
        first.Enqueue(DateTimeOffset.UtcNow);
        first.Enqueue(DateTimeOffset.UtcNow);

        var second = new ActivityQueue(_tempFile);
        Assert.Equal(2, second.PendingCount);
    }

    [Fact]
    public async Task AfterSuccessfulFlush_AFreshInstanceHasNothingQueued()
    {
        var first = new ActivityQueue(_tempFile);
        first.Enqueue(DateTimeOffset.UtcNow);
        await first.FlushAsync(new FakeEventsApiClient(), "https://example.vercel.app", "k");

        var second = new ActivityQueue(_tempFile);
        Assert.Equal(0, second.PendingCount);
    }

    [Fact]
    public void LastSuccessfulSyncAt_IsNullInitially()
    {
        var queue = new ActivityQueue(_tempFile);
        Assert.Null(queue.LastSuccessfulSyncAt);
    }

    [Fact]
    public async Task Flush_OnSuccess_SetsLastSuccessfulSyncAt()
    {
        var now = new DateTimeOffset(2026, 1, 1, 12, 0, 0, TimeSpan.Zero);
        var queue = new ActivityQueue(_tempFile, () => now);
        queue.Enqueue(now);

        await queue.FlushAsync(new FakeEventsApiClient(), "https://example.vercel.app", "k");

        Assert.Equal(now, queue.LastSuccessfulSyncAt);
    }

    [Fact]
    public async Task Flush_OnFailure_DoesNotSetLastSuccessfulSyncAt()
    {
        var queue = new ActivityQueue(_tempFile);
        queue.Enqueue(DateTimeOffset.UtcNow);
        var client = new FakeEventsApiClient { ShouldFail = true };

        await queue.FlushAsync(client, "https://example.vercel.app", "k");

        Assert.Null(queue.LastSuccessfulSyncAt);
    }

    [Fact]
    public async Task LastSuccessfulSyncAt_SurvivesRestart_ByPersistingToDisk()
    {
        var now = new DateTimeOffset(2026, 1, 1, 12, 0, 0, TimeSpan.Zero);
        var first = new ActivityQueue(_tempFile, () => now);
        first.Enqueue(now);
        await first.FlushAsync(new FakeEventsApiClient(), "https://example.vercel.app", "k");

        var second = new ActivityQueue(_tempFile);
        Assert.Equal(now, second.LastSuccessfulSyncAt);
    }

    [Fact]
    public void LoadsAnOlderQueueFile_WrittenBeforeLastSuccessfulSyncAtExisted()
    {
        // Pre-upgrade queue files are a bare `[String]` of ISO 8601
        // timestamps, with no lastSuccessfulSyncAt field at all.
        var directory = Path.GetDirectoryName(_tempFile)!;
        Directory.CreateDirectory(directory);
        var legacyJson = System.Text.Json.JsonSerializer.Serialize(new[] { DateTimeOffset.UtcNow.ToString("yyyy-MM-ddTHH:mm:ss.fffZ") });
        File.WriteAllText(_tempFile, legacyJson);

        var queue = new ActivityQueue(_tempFile);

        Assert.Equal(1, queue.PendingCount);
        Assert.Null(queue.LastSuccessfulSyncAt);
    }

    // ---- Outage resilience ----

    private const string Url = "https://example.vercel.app";
    private static DateTimeOffset T(int seconds) => DateTimeOffset.FromUnixTimeSeconds(seconds);

    [Fact]
    public async Task Flush_SendsLargeBacklogInChunks_AndClearsAll()
    {
        var queue = new ActivityQueue(_tempFile, chunkSize: 100, persistDebounce: TimeSpan.Zero);
        for (var i = 0; i < 250; i++) queue.Enqueue(T(i));
        var client = new FakeEventsApiClient();

        await queue.FlushAsync(client, Url, "k");

        Assert.Equal(new[] { 100, 100, 50 }, client.ReceivedBatches.Select(b => b.Count));
        Assert.Equal(0, queue.PendingCount);
    }

    [Fact]
    public async Task Flush_FailureMidBacklog_KeepsOnlyUnsentChunks()
    {
        var queue = new ActivityQueue(_tempFile, chunkSize: 100, persistDebounce: TimeSpan.Zero);
        for (var i = 0; i < 250; i++) queue.Enqueue(T(i));
        var client = new FakeEventsApiClient();
        var calls = 0;
        client.OnPost = () => client.ShouldFail = ++calls >= 2;

        await queue.FlushAsync(client, Url, "k");

        Assert.Equal(150, queue.PendingCount);
        Assert.Equal(T(100), queue.PendingTimestamps[0]);
    }

    [Fact]
    public async Task Flush_KeepsEventsEnqueuedWhileRequestIsInFlight()
    {
        var queue = new ActivityQueue(_tempFile, persistDebounce: TimeSpan.Zero);
        queue.Enqueue(T(1));
        var client = new FakeEventsApiClient();
        var injected = false;
        client.OnPost = () =>
        {
            if (injected) return;
            injected = true;
            queue.Enqueue(T(2));
        };

        await queue.FlushAsync(client, Url, "k");

        Assert.Equal(new[] { T(1) }, client.ReceivedBatches[0]);
        Assert.Equal(2, client.ReceivedBatches.Sum(b => b.Count)); // late arrival sent by the next chunk, not lost
        Assert.Equal(0, queue.PendingCount);
    }

    [Fact]
    public async Task Flush_CapEvictionDuringFlight_DoesNotRemoveUnsentEntries()
    {
        var queue = new ActivityQueue(_tempFile, maxPendingCount: 3, persistDebounce: TimeSpan.Zero);
        for (var i = 0; i < 3; i++) queue.Enqueue(T(i));
        var client = new FakeEventsApiClient();
        var first = true;
        client.OnPost = () =>
        {
            if (!first) return;
            first = false;
            queue.Enqueue(T(3)); // evicts 0
            queue.Enqueue(T(4)); // evicts 1
        };

        await queue.FlushAsync(client, Url, "k");

        Assert.Equal(0, queue.PendingCount);
        Assert.Equal(T(4), client.ReceivedBatches.SelectMany(b => b).Last());
    }

    [Fact]
    public async Task Flush_PermanentRejection_DropsChunkAndContinues()
    {
        var queue = new ActivityQueue(_tempFile, chunkSize: 2, persistDebounce: TimeSpan.Zero);
        for (var i = 0; i < 4; i++) queue.Enqueue(T(i));
        var client = new FakeEventsApiClient { Failure = new ApiClientException("bad", 400) };
        var calls = 0;
        client.OnPost = () => client.ShouldFail = ++calls == 1;

        await queue.FlushAsync(client, Url, "k");

        Assert.Equal(0, queue.PendingCount);
        Assert.Equal(2, queue.DroppedEventCount);
        Assert.Equal(2, client.ReceivedBatches.Count);
    }

    [Fact]
    public async Task Flush_Unauthorized_KeepsDataBacksOffAndReportsError()
    {
        var queue = new ActivityQueue(_tempFile);
        queue.Enqueue(T(1));
        var client = new FakeEventsApiClient { ShouldFail = true, Failure = new ApiClientException("nope", 401) };

        await queue.FlushAsync(client, Url, "k");

        Assert.Equal(1, queue.PendingCount);
        Assert.Equal("API key invalid or revoked", queue.LastError);
        Assert.Equal(1, queue.ConsecutiveFailureCount);
    }

    [Fact]
    public async Task Flush_AfterFailure_SkipsRetryUntilBackoffElapses()
    {
        var now = new DateTimeOffset(2026, 1, 1, 12, 0, 0, TimeSpan.Zero);
        var queue = new ActivityQueue(_tempFile, () => now, baseBackoff: TimeSpan.FromSeconds(10));
        queue.Enqueue(now);
        var client = new FakeEventsApiClient { ShouldFail = true };

        await queue.FlushAsync(client, Url, "k");
        now += TimeSpan.FromSeconds(5);
        await queue.FlushAsync(client, Url, "k");
        Assert.Single(client.ReceivedBatches);

        now += TimeSpan.FromSeconds(6);
        await queue.FlushAsync(client, Url, "k");
        Assert.Equal(2, client.ReceivedBatches.Count);
        Assert.Equal(2, queue.ConsecutiveFailureCount);
    }

    [Fact]
    public void BackoffInterval_DoublesUpToCap()
    {
        var b = TimeSpan.FromSeconds(10);
        var max = TimeSpan.FromSeconds(100);
        Assert.Equal(TimeSpan.FromSeconds(10), ActivityQueue.BackoffInterval(1, b, max));
        Assert.Equal(TimeSpan.FromSeconds(40), ActivityQueue.BackoffInterval(3, b, max));
        Assert.Equal(TimeSpan.FromSeconds(100), ActivityQueue.BackoffInterval(6, b, max));
    }

    [Fact]
    public async Task Flush_Success_ClearsLastError()
    {
        var now = new DateTimeOffset(2026, 1, 1, 12, 0, 0, TimeSpan.Zero);
        var queue = new ActivityQueue(_tempFile, () => now, baseBackoff: TimeSpan.FromSeconds(10));
        queue.Enqueue(now);
        var client = new FakeEventsApiClient { ShouldFail = true };
        await queue.FlushAsync(client, Url, "k");
        Assert.NotNull(queue.LastError);

        client.ShouldFail = false;
        now += TimeSpan.FromSeconds(20);
        await queue.FlushAsync(client, Url, "k");

        Assert.Null(queue.LastError);
    }

    [Fact]
    public void Enqueue_BeyondCap_DropsOldestAndCounts()
    {
        var queue = new ActivityQueue(_tempFile, maxPendingCount: 2, persistDebounce: TimeSpan.Zero);
        for (var i = 0; i < 5; i++) queue.Enqueue(T(i));

        Assert.Equal(new[] { T(3), T(4) }, queue.PendingTimestamps);
        Assert.Equal(3, queue.DroppedEventCount);
    }

    [Fact]
    public void Enqueue_WithinDebounceWindow_IsOnlyPersistedByPersistNow()
    {
        var now = new DateTimeOffset(2026, 1, 1, 12, 0, 0, TimeSpan.Zero);
        var first = new ActivityQueue(_tempFile, () => now, persistDebounce: TimeSpan.FromSeconds(5));
        first.Enqueue(T(1));
        now += TimeSpan.FromSeconds(1);
        first.Enqueue(T(2));
        Assert.Equal(1, new ActivityQueue(_tempFile).PendingCount);

        first.PersistNow();
        Assert.Equal(2, new ActivityQueue(_tempFile).PendingCount);
    }

    [Fact]
    public void Load_CorruptQueueFile_IsQuarantinedNotOverwritten()
    {
        Directory.CreateDirectory(Path.GetDirectoryName(_tempFile)!);
        File.WriteAllText(_tempFile, "{not json");

        var queue = new ActivityQueue(_tempFile, persistDebounce: TimeSpan.Zero);
        queue.Enqueue(T(1));

        Assert.Equal(1, queue.PendingCount);
        Assert.Equal("{not json", File.ReadAllText(_tempFile + ".corrupt"));
    }

    // ---- Backward compatibility with queue files written by the previous version ----

    [Fact]
    public async Task ExistingQueueFileWith3000Events_LoadsAndIsFullyDeliveredInChunks()
    {
        // Exactly the shape the previous version wrote: an object with
        // "pending" and "lastSuccessfulSyncAt", ms-precision Z timestamps.
        Directory.CreateDirectory(Path.GetDirectoryName(_tempFile)!);
        var stamps = Enumerable.Range(0, 3000).Select(i => T(1_700_000_000 + i * 30).UtcDateTime.ToString("yyyy-MM-ddTHH:mm:ss.fffZ")).ToArray();
        File.WriteAllText(_tempFile, System.Text.Json.JsonSerializer.Serialize(new { pending = stamps, lastSuccessfulSyncAt = "2026-01-01T00:00:00.000Z" }));

        var queue = new ActivityQueue(_tempFile);
        Assert.Equal(3000, queue.PendingCount);
        Assert.NotNull(queue.LastSuccessfulSyncAt);

        var client = new FakeEventsApiClient();
        await queue.FlushAsync(client, Url, "k");

        Assert.Equal(3000, client.ReceivedBatches.Sum(b => b.Count));
        Assert.All(client.ReceivedBatches, b => Assert.True(b.Count <= ActivityQueue.DefaultChunkSize));
        Assert.Equal(0, queue.PendingCount);
    }
}
